### PR TITLE
Fixes to select and ngOptions - "required" validation and behavior with empty / unknown option

### PR DESCRIPTION
**Pull request by Narretz**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
bugfixes

**What is the current behavior? (You can also link to an open issue here)**
- (a) when the unknown option is selected and the select is "required", the required error is not set
- (b) ngOptions: when the model is unmatched and an empty option exists, the empty option is always selected, even if the model is not null / undefined.

**What is the new behavior (if this is a feature change)?**
- (a) the error is set
- (b) the unknown option is selected if the model is not null / undefined

**Does this PR introduce a breaking change?**
Possibly? For (a), the following situation is possible: The model is programatically set to a value that is no longer in select options, and the select is required. With this patch, the user cannot keep the unmatched value, because of the required error, which considers this value invalid (just realized this, so it's not in the commit message).

For (b), it's unlikely that anyone relied on the fact that the empty option is selected instead of the unmatched option - the fixd behavior was also present in regular select since the beginning.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:

### CODE REVIEW

File: src/ng/directive/ngOptions.js [hunk 1]
```diff
@@ -449,12 +449,12 @@ var ngOptionsDirective = ['$compile', '$document', '$parse', function($compile,
       if (!multiple) {
 
         selectCtrl.writeValue = function writeNgOptionsValue(value) {
-          var selectedOption = options.selectValueMap[selectElement.val()];
+          var selectedOption = selectElement[0].options[selectElement[0].selectedIndex];
```
Comment (mgol) [inline src/ng/directive/ngOptions.js]: Why do you need to skip jqLite/jQuery here?

**Comment by petebacondarwin** [comment]
I am not sure that (a) is the desired behaviour here. For me required error implies that the input is missing a value. In this scenario it has a value but it is an invalid one. It feels like saying that a number input box has the required error if its value is higher than the maximum allowed.

**Comment by Narretz** [comment]
@petebacondarwin Hmm ... I implemented this based on the assumptions / opinions found here: https://github.com/angular/angular.js/issues/13172 I know, it's been a while :D

I agree that "unknown" is different than "empty" - the latter specifically means the select is "unselected", so setting required error seems solid. For unknown, I guess it depends on the app logic if it fulfills the required condition. The thing is that selects basically only have required validation, because the app is usually responsible for giving the select list.

I don't insist on this change, but I think we should provide a way to manipulate this behavior. The custom isEmpty method can be implemented by a 3rd party directive, but for that we need to expose / document properties of the select controller, mainly ctrl.unknownOption ... and maybe some others. This would obviously put us in a position where we expose more API that we have to maintain :o But it would also allow people to create a different error, e.g. "unmatched value" that is different than  "required".

File: src/ng/directive/select.js [hunk 2]
```diff
@@ -650,6 +655,21 @@ var selectDirective = function() {
...
+
+        // A single select is also considered empty when the unknown option is present, or when
+        // the empty option is selected
+        ngModelCtrl.$isEmpty = function(value) {
+
+          return originalIsEmpty(value) ||
+            !!selectCtrl.unknownOption.parent().length ||
+            // When the empty option is selected from the scope, then the viewValue can be different
+            // than '', and the originalIsEmpty fn will not catch this case
+            (selectCtrl.hasEmptyOption &&
```
Comment (Narretz) [inline src/ng/directive/select.js]: Note that this condition only affects empty option required validation, not unknown option - commit message needs to be updated.

File: src/ng/directive/ngOptions.js (hunk 1 above)
Comment (Narretz) [inline src/ng/directive/ngOptions.js]: We don't. I just didn't think to use jQuery here. How do you get the selected option with jquery?

Comment (mgol) [inline src/ng/directive/ngOptions.js]: Hmm, I was just comparing it to the previous code which used `options.selectValueMap`. Could you explain the change?

**Comment by petebacondarwin** [comment]
OK, I think exposing some more public API to empower developers is a reasonable goal here.

File: src/ng/directive/ngOptions.js (hunk 1 above)
Comment (Narretz) [inline src/ng/directive/ngOptions.js]: Ah, I see. Regular select was using the same technique - getting the option directly from the DOM, so I made it consistent. I don't think there's a huge difference, it just felt right to look at the actual DOM

**Comment by Narretz** [comment]
I've removed the change about required and the unknown option, and expanded on the tests around this behavior. PR is ready to review.
Select API changes will follow in a different PR.

**Comment by petebacondarwin** [comment]
I'll review tomorrow morning

**Review by petebacondarwin — COMMENTED** [review]
Nice work @Narretz - LGTM

**Review by petebacondarwin — APPROVED** [review]

**Comment by petebacondarwin** [comment]
Please merge when you have a moment @Narretz

**Pull request merged**